tensorflow/lite/kernels/internal/optimized/optimized_ops.h

```diff
@@ -5277,56 +5277,66 @@ inline void PadImageStyle(const tflite::PadParams& op_params,
-  const int stop_b = (4 - size_count > 0 || op_params.size[0] == -1)
-                         ? ext_shape.Dims(0)
-                         : start_b + op_params.size[0];
-  const int start_h = begin_count < 3 ? 0 : op_params.begin[begin_count - 3];
-  const int stop_h = (size_count < 3 || op_params.size[size_count - 3] == -1)
-                         ? ext_shape.Dims(1)
-                         : start_h + op_params.size[size_count - 3];
-  const int start_w = begin_count < 2 ? 0 : op_params.begin[begin_count - 2];
-  const int stop_w = (size_count < 2 || op_params.size[size_count - 2] == -1)
-                         ? ext_shape.Dims(2)
-                         : start_w + op_params.size[size_count - 2];
-  const int start_d = begin_count < 1 ? 0 : op_params.begin[begin_count - 1];
-  const int stop_d = (size_count < 1 || op_params.size[size_count - 1] == -1)
-                         ? ext_shape.Dims(3)
-                         : start_d + op_params.size[size_count - 1];
-
-  for (int in_b = start_b; in_b < stop_b; ++in_b) {
-    for (int in_h = start_h; in_h < stop_h; ++in_h) {
-      for (int in_w = start_w; in_w < stop_w; ++in_w) {
-        const int len = stop_d - start_d;
-        if (len > 0)
-          writer->WriteN(Offset(ext_shape, in_b, in_h, in_w, start_d), len);
-      }
-    }
-  }
+  const int slice_dimensions = input_shape.DimensionsCount();
+
+  // Below values are under condition if the input is 1-D, as it is a special
+  // case, where we will not have next axis to fetch the copy length which is
+  // input for memcpy operation below, so we are alligning the variables to the
+  // required in the case where input is 1-D. NOTE: This is done only to attain
+  // code reuse of compute_slice() defined below.
+  bool is_input_1d = slice_dimensions == 1;
+  int final_axis = is_input_1d ? 0 : slice_dimensions - 2;
+  int copy_len =
+      is_input_1d
+          ? op_params.size[final_axis] - op_params.begin[final_axis]
+          : op_params.size[final_axis + 1] - op_params.begin[final_axis + 1];
+  int dim_mul = is_input_1d ? 1 : input_shape.Dims(final_axis + 1);
+  int start_fin = is_input_1d ? 0 : op_params.begin[final_axis + 1];
+
+  if (is_input_1d) {
+    // (This is very important): Set size so that the final loop will run only
+    // once to accommodate complete memcpy operation
+    op_params.size[final_axis] = op_params.begin[final_axis] + 1;
+  }
+
+  // Store as std::function to allow recursion.
+  std::function<void(int, int)> compute_slice =
+      [&compute_slice, final_axis, copy_len, dim_mul, start_fin, writer,
+       input_shape, op_params](int axis, int curr_indices) {
+        if (axis == (final_axis)) {
+          for (int in_x = op_params.begin[axis]; in_x < op_params.size[axis];
+               in_x++) {
+            int index = (in_x + curr_indices) * dim_mul + start_fin;
+            if (copy_len > 0)
+              writer->WriteN(index, copy_len);
```

Comment:
Is WriteN the only real optimization here? If so, I'd be more than OK using WriteN in the reference impl, and removing this version.

Reply:
@jdduke : The main difference is in Reference(It will iterate through all axis, then fill the output buffer), but in Optimized(It will iterate only axis - 1, and last axis it will just copy the length directly to output buffer).
Please let me know, if still want to merge to one into Reference.
I also feel, it should be moved to reference.